zeth-contracts/contracts/BaseMixer.sol

```diff
@@ -130,73 +216,38 @@ contract BaseMixer is MerkleTreeMiMC7, ERC223ReceivingContract {
-
-        //Format and append the nullifiers
-        for(uint i = 1; i < 1 + 2 * (jsIn); i += 2) {
-            digest_inputs[0] = primary_inputs[i];
-            digest_inputs[1] = primary_inputs[i+1];
-            formatted = Bytes.sha256_digest_from_field_elements(digest_inputs);
-            formatted_inputs[(i-1)/2 + 1] = formatted;
-        }
-
-        //Format and append the commitments
-        for(uint i = 1 + 2 * (jsIn); i < 1 + 2 * (jsIn + jsOut); i += 2) {
-            digest_inputs[0] = primary_inputs[i];
-            digest_inputs[1] = primary_inputs[i+1];
-            formatted = Bytes.sha256_digest_from_field_elements(digest_inputs);
-            formatted_inputs[(i-1)/2 + 1] = formatted;
-        }
-
-        //Format and append the v_pub_in
-        formatted = bytes32(primary_inputs[1 + 2 * (jsIn + jsOut)]);
-        formatted_inputs[1 + jsIn + jsOut] = formatted;
-
-        //Format and append the v_pub_out
-        formatted = bytes32(primary_inputs[1 + 2 * (jsIn + jsOut) + 1]);
-        formatted_inputs[1 + jsIn + jsOut + 1] = formatted;
-
-        //Format and append h_sig
-        digest_inputs[0] = primary_inputs[1 + 2 * (jsIn + jsOut) + 1 + 1];
-        digest_inputs[1] = primary_inputs[1 + 2 * (jsIn + jsOut + 1) + 1];
-        formatted = Bytes.sha256_digest_from_field_elements(digest_inputs);
-        formatted_inputs[1 + jsIn + jsOut + 1 + 1] = formatted;
-
-        //Format and append the h_iS
-        for(uint i = 1 + 2 * (jsIn + jsOut + 1) + 1 + 1; i < 1 + 2 * (jsIn + jsOut + 1 + jsIn) + 1 + 1; i += 2) {
-            digest_inputs[0] = primary_inputs[i];
-            digest_inputs[1] = primary_inputs[i+1];
-            formatted = Bytes.sha256_digest_from_field_elements(digest_inputs);
-            formatted_inputs[(i-1)/2 + 2] = formatted;
-        }
-
-        bytes32 hash_inputs = sha256(abi.encodePacked(formatted_inputs));
-
-        return hash_inputs;
-    }
-
-
     function assemble_commitments_and_append_to_state(uint[] memory primary_inputs) internal {
         // We re-assemble the commitments (JSOutputs)
-        uint256[] memory digest_inputs = new uint[](2);
-        for(uint i = 1 + 2 * jsIn ; i < 1 + 2*(jsIn + jsOut); i += 2) {
-            digest_inputs[0] = primary_inputs[i]; // See the way the inputs are ordered in the extended proof
-            digest_inputs[1] = primary_inputs[i+1];
-            bytes32 current_commitment = Bytes.sha256_digest_from_field_elements(digest_inputs);
+        uint index;
+        bytes1 bits_input;
+        uint size_extra_bits = digest_length % field_capacity;
+        if (digest_length < field_capacity) {
+            size_extra_bits = 0;
+        }
+        for(uint i = 1 + jsIn ; i < 1 + jsIn + jsOut; i ++) {
+            // See the way the inputs are ordered in the extended proof
+            index = (digest_length-field_capacity)*(i-1);
+            bits_input = extract_extra_bits(index, index+(size_extra_bits-1), primary_inputs);
```

Comment:
I used the modulo (`uint size_extra_bits = digest_length % field_capacity;`) as a generalisation (say we have `digest_length = 2*field_capacity + a` for some `a`).

The `-1` is not to extract the last position asked. I can see how this leads to a problem. A solution would be to change the function definitions so that the second argument is the number of bits we want to extract and not the position of the last bit we want to extract.

I wrote `size_extra_bits = 0` to have an edge case where we do not need to extract any bits. The `-1` however brings an error (we'll have the start position > end position, and potentially the end position equal to `-1` resulting in an overflow). The previous solution should resolve it.

Reply:
> I did not use modulo consistently however which may lead to problems (if a simple substraction is not enough).

Yes

> A solution would be to change the function definitions so that the second argument is the number of bits we want to extract and not the position of the last bit we want to extract

I like this solution better yes

Reply:
Fixed.